File: frontend/lib/api/user.service.ts
import { apiClient } from './client';
import {
  User,
  CreateUserRequest,
  UpdateUserRequest,
  TransactionResponse,
  PaginatedResponse,
  UploadPictureRequest,
  UploadPictureResponse,
  UserResponse,
} from '@/lib/types';

const ENDPOINTS = {
  USERS: '/api/users',
} as const;

export const userService = {
  getMe(): Promise<User> {
    return apiClient.get(`${ENDPOINTS.USERS}/me`);
  },

  getAllUsers(): Promise<User[]> {
    return apiClient.get(ENDPOINTS.USERS);
  },

  createUser(data: CreateUserRequest): Promise<User> {
    return apiClient.post(ENDPOINTS.USERS, data);
  },

  getUserById(id: string): Promise<User> {
    return apiClient.get(`${ENDPOINTS.USERS}/${id}`);
  },

  getMyProfile(): Promise<User> {
    return apiClient.get(`${ENDPOINTS.USERS}/me`);
  },

  updateProfilePicture(data: UploadPictureRequest): Promise<UploadPictureResponse> {
    return apiClient.put(`${ENDPOINTS.USERS}/me/profile-picture`, data);
  },

  getUserTransactions(
    id: string,
    params?: { page?: number; size?: number; sortBy?: string; sortDirection?: string },
  ): Promise<PaginatedResponse<TransactionResponse>> {
    const query = new URLSearchParams();
    if (params?.page !== undefined) query.set('page', String(params.page));
    if (params?.size !== undefined) query.set('size', String(params.size));
    if (params?.sortBy) query.set('sortBy', params.sortBy);
    if (params?.sortDirection) query.set('sortDirection', params.sortDirection);
    const qs = query.toString();
    return apiClient.get(`${ENDPOINTS.USERS}/${id}/transactions${qs ? `?${qs}` : ''}`);
  },

  getMyTransactions(
    params?: { page?: number; size?: number; sortBy?: string; sortDirection?: string },
  ): Promise<PaginatedResponse<TransactionResponse>> {
    const query = new URLSearchParams();
    if (params?.page !== undefined) query.set('page', String(params.page));
    if (params?.size !== undefined) query.set('size', String(params.size));
    if (params?.sortBy) query.set('sortBy', params.sortBy);
    if (params?.sortDirection) query.set('sortDirection', params.sortDirection);
    const qs = query.toString();
    return apiClient.get(`${ENDPOINTS.USERS}/me/transactions${qs ? `?${qs}` : ''}`);
  },

  getUserFriends(
    id: string | number,
    params?: { page?: number; size?: number },
  ): Promise<PaginatedResponse<User>> {
    const query = new URLSearchParams();
    if (params?.page !== undefined) query.set('page', String(params.page));
    if (params?.size !== undefined) query.set('size', String(params.size));
    const qs = query.toString();
    return apiClient.get(`${ENDPOINTS.USERS}/${id}/friends${qs ? `?${qs}` : ''}`);
  },

  updateUser(id: string, data: UpdateUserRequest): Promise<User> {
    return apiClient.put(`${ENDPOINTS.USERS}/${id}`, data);
  },

  deleteUser(id: string): Promise<void> {
    return apiClient.delete(`${ENDPOINTS.USERS}/${id}`);
  },

  searchUsers(
    query: string,
    params?: { page?: number; size?: number }
  ): Promise<PaginatedResponse<UserResponse>> {
    const qs = new URLSearchParams();
    qs.set('query', query);
    if (params?.page !== undefined) qs.set('page', String(params.page));
    if (params?.size !== undefined) qs.set('size', String(params.size));
    return apiClient.get(`${ENDPOINTS.USERS}/search?${qs.toString()}`);
  },
};

File: backend/src/main/java/com/api/betdobem/services/UserService.java
package com.api.betdobem.services;

import com.api.betdobem.domain.User;
import com.api.betdobem.dtos.requests.CreateUserRequest;
import com.api.betdobem.dtos.requests.UpdatePictureRequest;
import com.api.betdobem.dtos.requests.UpdateUserRequest;
import com.api.betdobem.dtos.responses.*;
import com.api.betdobem.enums.UserRole;
import com.api.betdobem.mappers.UserMapper;
import com.api.betdobem.repositories.*;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Set;

@Service
public class UserService {
    private UserRepository userRepository;
    private UserMapper userMapper;
    private WalletService walletService;
    private S3StorageService s3StorageService;
    private BetRepository betRepository;
    private ActivityRepository activityRepository;
    private VoteRepository voteRepository;
    private FriendshipRepository friendshipRepository;

    public UserService(UserRepository userRepository, UserMapper userMapper, WalletService walletService, S3StorageService s3StorageService, BetRepository betRepository, ActivityRepository activityRepository, VoteRepository voteRepository, FriendshipRepository friendshipRepository) {
        this.userRepository = userRepository;
        this.userMapper = userMapper;
        this.walletService = walletService;
        this.s3StorageService = s3StorageService;
        this.betRepository = betRepository;
        this.activityRepository = activityRepository;
        this.voteRepository = voteRepository;
        this.friendshipRepository = friendshipRepository;
    }

    public List<UserResponse> getAllUsers() {
        List<User> users = userRepository.findAll();
        return userMapper.toUserResponseList(users);
    }

    public User getUserEntityById(Long id) {
        return userRepository.findById(id).orElseThrow(() -> new EntityNotFoundException("User with ID " + id + " not found."));
    }

    public List<User> getUserEntitiesByIds(Set<Long> ids) {
        return userRepository.findAllById(ids);
    }

    public User getUserEntityByEmail(String email) {
        return userRepository.findByEmail(email).orElseThrow(() -> new EntityNotFoundException("User with email " + email + " not found."));
    }

    public User getUserEntityByUsername(String username) {
        return userRepository.findByUsername(username).orElseThrow(() -> new EntityNotFoundException("User with username " + username + " not found."));
    }

    public boolean existsUserByEmail(String email) {
        return userRepository.existsByEmail(email);
    }

    public boolean existsUserByUsername(String username) {
        return userRepository.existsByUsername(username);
    }

    public PagedResponse<TransactionResponse> getUserTransactions(Long id, Pageable pageable) {
        return walletService.getUserTransactions(id, pageable);
    }

    public UserResponse createUser(CreateUserRequest user) {
        User userEntity = userMapper.toUserEntity(user);
        userEntity.setRole(UserRole.USER);
        User savedUser = userRepository.save(userEntity);
        return userMapper.toUserResponse(savedUser);
    }

    public UserResponse getUserById(Long id) {
        User user = getUserEntityById(id);
        return userMapper.toUserResponse(user);
    }

    public UserProfileResponse getUserProfileById(Long id) {
        User user = getUserEntityById(id);
        long winningBets = betRepository.countWinningBetsByUserId(id);
        long registeredActivities = activityRepository.countByAuthorId(id);
        long computedVotes = voteRepository.countByVoterId(id);
        long friendsCount = friendshipRepository.countByUserIdOrFriendId(id, id);
        return userMapper.toUserProfileResponse(user, winningBets, registeredActivities, computedVotes, friendsCount);
    }

    public PagedResponse<FriendResponse> getUserFriends(Long id, Pageable pageable) {
        User user = getUserEntityById(id);
        Page<User> friends = friendshipRepository.findFriendsByUserId(user.getId(), pageable);
        List<FriendResponse> friendsList = userMapper.toFriendResponseList(friends.getContent());
        return new PagedResponse<>(
                friendsList,
                friends.getNumber(),
                friends.getSize(),
                friends.getTotalElements(),
                friends.getTotalPages(),
                friends.hasNext()
        );
    }

    public UserResponse updateUser(Long id, UpdateUserRequest user) {
        User existingUser = getUserEntityById(id);
        userMapper.updateUserRequest(user, existingUser);
        User updatedUser = userRepository.save(existingUser);
        return userMapper.toUserResponse(updatedUser);
    }

    public UploadPictureResponse setProfilePicture(UpdatePictureRequest picture, Long id) {
        User user = getUserEntityById(id);
        String uniqueObjectKey = String.format("users/profiles/user_%d_%d_%s",
                id,
                System.currentTimeMillis(),
                picture.fileName().replaceAll("[^a-zA-Z0-9.-]", "_"));
        user.setProfilePictureUrl(uniqueObjectKey);
        userRepository.save(user);
        String uploadUrl = s3StorageService.generatePresignedUploadUrl(uniqueObjectKey, picture.contentType());
        return new UploadPictureResponse(uploadUrl);
    }

    public void deleteUser(Long id) {
        userRepository.deleteById(id);
    }

    public PagedResponse<UserResponse> searchUsers(String query, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        Page<User> usersPage = userRepository.searchUsers(query, pageable);
        return new PagedResponse<>(
                userMapper.toUserResponseList(usersPage.getContent()),
                usersPage.getNumber(),
                usersPage.getSize(),
                usersPage.getTotalElements(),
                usersPage.getTotalPages(),
                usersPage.hasNext()
        );
    }
}
